### ISSUE TITLE
Memory corruption issue (0.1.4)

### ISSUE THREAD

**Issue by jpf91** [thaven/oauth#5]
@thaven @wilzbach 
Seems like this is yet another victim of issue https://issues.dlang.org/show_bug.cgi?id=12625 .

In 0.1.4 there's this line in `client.d`:
```d
this.hash = sha256Of(provider.tokenUri ~ ' ' ~ clientId);
```

This assigns stack allocated memory to this.hash and the value i later overwritten by other stack data. Looks like this code was refactored so maybe master is not affected. Nevertheless a bugfix release would be nice. Solution: Use `.idup`.

I also suggest to be careful with the `loginKey/LoginData.key` methods. AFAICS the code won't cause corruption, but as you return a static array from this function you have to make sure all uses of this function are safe (essentially you have to threat the returned array as `scope`).

**Comment by jpf91** [comment]
BTW: Could we also get a vibe 0.7.x version with the github provider backported? AFAICS it's a simple copy/paste:

```d
/++
    OAuth 2.0 for D - Settings for GitHub

    Copyright: © 2016,2017 Harry T. Vennik
    License: Subject to the terms of the MIT license, as written in the included LICENSE file.
    Authors: Sebastian Wilzbach
  +/

module oauth.provider.github;

import oauth.client : OAuthProvider;

shared static this()
{
    OAuthProvider.register("github", new immutable(OAuthProvider)(
        "https://github.com/login/oauth/authorize",
        "https://github.com/login/oauth/access_token"
    ));
}
```

**Comment by thaven** [comment]
Looks like it's been fixed in master: https://github.com/thaven/oauth/blob/eb2b0a74296cedd71c40a2a39041ec0c036532e4/source/oauth/settings.d#L139

But this should still be fixed in 0.1.x range too. I'll backport the fix sometime soon.

May indeed include the GitHub provider as well.

**Comment by wilzbach** [comment]
@jpf91: yep I ran into this issue as well (that's why it's fixed in master). It's pretty nasty because the compiler should yield an error here ...

**Comment by thaven** [comment]
Fixed in a665a9d1857d15be2463d10d62705e3f1b361e5b

**Comment by jpf91** [comment]
Awesome, now everything works perfectly (github App using `jwtd` + github outh provider).
Thanks a lot!
